add `window.originAgentCluster` (#30008)
mdn/content: files/en-us/web/api/window/originagentcluster/index.md

---
title: "Window: originAgentCluster property"
short-title: originAgentCluster
slug: Web/API/Window/originAgentCluster
page-type: web-api-instance-property
browser-compat: api.Window.originAgentCluster
---

{{APIRef}}{{SeeCompatTable}}

The **`originAgentCluster`** read-only property of the {{domxref("Window")}} interface returns `true` if this window belongs to an agent cluster (a maximal set of agents that can communicate by operating on shared memory, every agent belongs to exactly one agent cluster) which is origin-keyed.

> A Document delivered over a secure context can request that it be placed in an origin-keyed agent cluster, by using the `Origin-Agent-Cluster` HTTP response header. The consequences are that the resulting Document's agent cluster key is its origin, instead of the corresponding site. In terms of observable effects, this means that attempting to relax the [same-origin](/en-US/docs/Web/Security/Same-origin_policy) restriction using `document.domain` will instead do nothing, and it will not be possible to send WebAssembly.Module objects to [cross-origin](/en-US/docs/Web/HTTP/CORS) Documents (even if they are same site). Behind the scenes, this isolation can allow user agents to allocate implementation-specific resources corresponding to agent clusters, such as processes or threads, more efficiently.

Reviewer: This paragraph is almost an exact copy from the [specification](https://html.spec.whatwg.org/multipage/browsers.html#origin-keyed-agent-clusters). Please don't do this: it's our job as MDN writers to make the specs more accessible to web developers. If we're not doing that we're not really adding value above the spec. Also, the spec uses many technical terms with very precise definitions, and always links to them. For example, "corresponding site" here. When we just copy spec text we lose these links, so the text no longer makes sense.

The specs are primarily aimed at browser developers. When we write content for MDN we have to think about how we will explain the technology to web developers, and take care to provide as much context as we need to, for our explanations to make sense.

---
title: "Window: originAgentCluster property"
short-title: originAgentCluster
slug: Web/API/Window/originAgentCluster
page-type: web-api-instance-property
browser-compat: api.Window.originAgentCluster
---

{{APIRef}}{{SeeCompatTable}}

> The **`originAgentCluster`** read-only property of the {{domxref("Window")}} interface returns `true` if this window belongs to an _origin-keyed_ [agent cluster](https://tc39.es/ecma262/#sec-agent-clusters), this means that the operating system has provided dedicated resources (for example an operating system process) to this window's origin that are not shared with windows from other origins. and `false` otherwise.

Reviewer: ```suggestion
The **`originAgentCluster`** read-only property of the {{domxref("Window")}} interface returns `true` if this window belongs to an _origin-keyed [agent cluster](https://tc39.es/ecma262/#sec-agent-clusters)_: this means that the operating system has provided dedicated resources (for example an operating system process) to this window's origin that are not shared with windows from other origins.

Otherwise this property returns `false`.
```

Windows that are part of an origin-keyed cluster are subjects to some additional restrictions, compared with windows that are not. In particular, they cannot:

- Set {{domxref("Document.domain")}}, which is a legacy feature that normally allows same-site cross-origin pages to synchronously access each other's DOM.
- Send [`WebAssembly.Module`](/en-US/docs/WebAssembly/JavaScript_interface/Module) objects to other same-site cross-origin pages via {{domxref("Window.postMessage()")}}.

> - Send {{jsxref("SharedArrayBuffer")}} or [`WebAssembly.Memory`](/en-US/docs/WebAssembly/JavaScript_interface/Memory) objects to other same-site cross-origin pages. Note that this feature is only supported in Chrome and will disable in future versions chrome, see [Intent to Ship: Restriction on SharedArrayBuffers](https://groups.google.com/a/chromium.org/g/blink-dev/c/1NKvbIj3dq4/m/Vgfisu5HAwAJ).

Reviewer: ```suggestion
- Send {{jsxref("SharedArrayBuffer")}} or [`WebAssembly.Memory`](/en-US/docs/WebAssembly/JavaScript_interface/Memory) objects to other same-site cross-origin pages.
```

(I don't think this detail is needed in this context)

> Note that this feature is only available in {{glossary("Secure Context")}}.

Reviewer: This should be communicated using the `{{SecureContext_Header}}` macro at the top of the page (after `{{SeeCompatTable}}`).

Author: Well, I don't think we should use `{{SecureContext_Header}}` here, because the property can still be accessed in non Secure Context (in http), simply in such situation, the property will always return `false`

The sentence maybe a little confusing, so I make an update

Author: Well, I don't think we should use `{{SecureContext_Header}}` here, because the property can still be accessed in non Secure Context (in http), simply in such situation, the property will always return `false`

The sentence maybe a little confusing, so I make an update

Reviewer: ah, I see, fair enough :).

---
title: "Window: originAgentCluster property"
short-title: originAgentCluster
slug: Web/API/Window/originAgentCluster
page-type: web-api-instance-property
browser-compat: api.Window.originAgentCluster
---

{{APIRef}}{{SeeCompatTable}}

The **`originAgentCluster`** read-only property of the {{domxref("Window")}} interface returns `true` if this window belongs to an _origin-keyed [agent cluster](https://tc39.es/ecma262/#sec-agent-clusters)_: this means that the operating system has provided dedicated resources (for example an operating system process) to this window's origin that are not shared with windows from other origins.

Otherwise this property returns `false`.

> Windows that are part of an origin-keyed cluster are subjects to some additional restrictions, compared with windows that are not. In particular, they cannot:

Author: ```suggestion
Windows that are part of an origin-keyed agent cluster are subjects to some additional restrictions, compared with windows that are not. In particular, they cannot:
```

- Set {{domxref("Document.domain")}}, which is a legacy feature that normally allows same-site cross-origin pages to synchronously access each other's DOM.
- Send [`WebAssembly.Module`](/en-US/docs/WebAssembly/JavaScript_interface/Module) objects to other same-site cross-origin pages via {{domxref("Window.postMessage()")}}.
- Send {{jsxref("SharedArrayBuffer")}} or [`WebAssembly.Memory`](/en-US/docs/WebAssembly/JavaScript_interface/Memory) objects to other same-site cross-origin pages.

> Note that this feature is only available in {{glossary("Secure Context")}}.

Author: ```suggestion
Note that origin-keyed agent cluster feature is only supported in {{glossary("Secure Context")}}. If a site is not a Secure Context, the `window.originAgentCluster` will always return `false`.
```

> Note that origin-keyed agent cluster feature is only supported in {{glossary("Secure Context")}}. If a site is not a Secure Context, the `window.originAgentCluster` will always return `false`.

Reviewer: ```suggestion
Note that the origin-keyed agent cluster feature is only supported in {{glossary("Secure Context", "secure contexts")}}. If a site is not a secure context, the `window.originAgentCluster` will always return `false`.
```
